### PR TITLE
Hotfix for batching in older jax versions

### PR DESCRIPTION
**Pull request by dpanici**
hotfix, add compatibility layer to sharding.update call as it is different api before jax 0.6.2

### CODE REVIEW

**Comment by unalmis** [comment]
some comments
1. If a full reversion is being done, should make a new PR that also includes the reversion of this reversion.
1. It sounds like this reversion would prevent use of DESC on JAX versions beyond 0.6.2. ?
2. The changes we backported from JAX (I think also include some JAX versions before 0.6.2) include JAX's bugfixes for various batching issues. As this PR intends to revert those fixes, this leads to a danger in shipping a version frankenstein of code. If the JAX version is one which assumes those are fixed and they aren't we would get subtle issues.
3. It is hard enough to convince developers of other dependencies to fix their bugs. Anything short of a creating a GitHub issue with the actual bugfix, or even more difficult: with a minimial working example, always gets ignored. That is my experience with some jax dependencies at least. I doubt anyone will take #1599 seriously if we mix jax versions for example. I am especially hesitant because JAX is still fixing batching issues that caused subtle bugs in DESC #1522 and maybe still affects #1894 in some way which took a very long time to isolate and delayed work.

so basically be careful

**Comment by dpanici** [comment]
> some comments
> 
> 1. If a full reversion is being done, should make a new PR that also includes the reversion of this reversion.

can you explain this part? I dont follow

**Review by unalmis — CHANGES_REQUESTED** [review]
> > so basically be careful
> 
> 
> 
> Fair, I need to think more. I welcome any suggestions
> we either need to
> 
> - revert the changes that caused these specific import errors or
> 
> - change our requirements to jax ==0.6.2

Here is how I would have done the original PR if the jax tests had failed there it.
There are 3 functions on master that we took  from the newest jax version.

If you want to preserve DESC ability to run old jax versions (even though these old jax versions have bugs that effect our stuff), and there is no simple fix, then it's safer to make new legacy batching functions for those 3 functions.

Call those legacy functions only when jax has an old version.
see the top of the desc.batching file to see how to check the jax version. 

So update the citation of those 3 functions to include the version you are taking the code from. 

```
if jax < compatible version:
    from jax import batch_abd_renainder
else:
    def batch_and....
        """..... jax version xxxx"""
         stuff
```

File: .github/workflows/jax_tests.yml [hunk 1]
```diff
@@ -14,15 +14,16 @@ jobs:
       matrix:
         jax-version: [0.4.24, 0.4.25, 0.4.26, 0.4.27, 0.4.28, 0.4.29,
                       0.4.30, 0.4.31, 0.4.33, 0.4.34, 0.4.35, 0.4.37,
-                      0.4.38, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.6.0]
+                      0.4.38, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.6.0, 0.6.1,
```
Comment (unalmis) [inline .github/workflows/jax_tests.yml]: some of 0.5. is broken (we block from install)

File: desc/batching.py [hunk 2]
```diff
@@ -41,90 +38,124 @@
 else:
     from jax import linear_util as lu
 
+try:  # noqa : C901
```
Comment (unalmis) [inline desc/batching.py]: why not just try to import the batch and remainder function from jax and if it is not there then you use the one from your favorite jax version that passes all the github CI tests?

File: .github/workflows/jax_tests.yml (hunk 1 above)
Comment (YigitElma) [reply .github/workflows/jax_tests.yml]: Can you remind me what the problem was exactly? Numerics or installation failure?

Comment (unalmis) [reply .github/workflows/jax_tests.yml]: the latter, see commit message of #1868

File: desc/batching.py (hunk 2 above)
Comment (YigitElma) [reply desc/batching.py]: @f0uriest and you were not ok with that previously (conversation is on slack)

Comment (unalmis) [reply desc/batching.py]: In that slack conversation

1. I said the point of having these function instead of importing from JAX is because the old versions of JAX that DESC claims to support don't have the function. Therefore, I had to rewrite an implementation of it in DESC. I chose the most recent JAX implementation because that had the least bugs. If you are already providing a fallback method to use when JAX lacks the feature, then you can just import the proper method from jax in the try block.

2. Rory's point was that you shouldn't import private stuff from JAX. I don't follow how that applies because to reimplement this in DESC you have to import even more private functions from JAX. See your import from jax._src.util for example.

Comment (YigitElma) [reply desc/batching.py]: Whoever convinced him before can deal with it; I just resolved the issue.

Comment (unalmis) [reply desc/batching.py]: Ok, then if the fallback method you chose passes all the jax tests, then I will make the update the use the jax method if the import is succesful.

Comment (unalmis) [reply desc/batching.py]: Looks like the fallback didn't work. I think it will be more likely to work with the method I suggested in the requested changes (which is same as the suggestion I made in this thread).

Comment (YigitElma) [reply desc/batching.py]: The failing ones fail due to something else (`rtol` argument to `jnp.pinv` which is added in 0.4.29 apparently), let's see the rest

Comment (unalmis) [reply desc/batching.py]: The ones above 0.4.29 fail as well, so the fallback should be taken from an older jax version.

File: .github/workflows/jax_tests.yml [hunk 3]
```diff
@@ -14,15 +14,16 @@ jobs:
       matrix:
         jax-version: [0.4.24, 0.4.25, 0.4.26, 0.4.27, 0.4.28, 0.4.29,
                       0.4.30, 0.4.31, 0.4.33, 0.4.34, 0.4.35, 0.4.37,
-                      0.4.38, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.6.0]
+                      0.4.38, 0.5.0, 0.5.3, 0.6.0, 0.6.1, 0.6.2]
```
Comment (f0uriest) [inline .github/workflows/jax_tests.yml]: while we're at it we might as well also test against newer versions (through 0.7.2) and bump the max version accordingly

Comment (unalmis) [reply .github/workflows/jax_tests.yml]: - are you going to block the 0.7.0 due to the performance regressions in those versions? think it was with diffrax
- should test after nuffts are added as dependency. If the nuffts don't work with 0.7.2 yet i'm just going to revert and block 0.7.2.


0.7.2 came out this week and JAX advertises breaking changes

**Comment by unalmis** [comment]
Thank you for removing 0.7.2 as i requested. As you can see, the bug reports for batching in 0.7.2 are now beginning. Look up issue 32028

**Review by unalmis — APPROVED** [review]

**Review by unalmis — APPROVED** [review]

File: desc/batching.py (hunk 2 above)
Comment (unalmis) [reply desc/batching.py]: done

File: requirements.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-jax >= 0.4.24, != 0.4.36, != 0.5.1, != 0.5.2, <= 0.6.2
+jax >= 0.4.29, != 0.4.36, != 0.5.1, != 0.5.2, <= 0.6.2
```
Comment (f0uriest) [inline requirements.txt:1]: Are we bumping the max version here and excluding 0.7.0?

**Comment by unalmis** [comment]
I try not to add workarounds because it would pollute our code base. From my memory, there's a series of bugs with FFT accuracy before 0.6.2 and AD with indexing multi-dim arrays.

**Comment by unalmis** [comment]
yes

**Review by unalmis — DISMISSED** [review]

File: tests/test_compute_funs.py [hunk 5]
```diff
@@ -1546,13 +1546,13 @@ def test_contravariant_basis_vectors_PEST(eq):
...
 @pytest.mark.slow
-@pytest.mark.parametrize("eq", [get("W7-X")])
+@pytest.mark.parametrize("eq", [get("precise_QA")])
 def test_PEST_derivative_math(eq):
     """Verify math to write PEST derivative quantities by redefining θ to θ_PEST."""
     from desc.compute import data_index
 
     tol = 1e-10
-    eq_PEST = eq.to_sfl(4 * eq.L, 5 * eq.M, 4 * eq.N, copy=True, tol=tol)
+    eq_PEST = eq.to_sfl(3 * eq.L, 3 * eq.M, 3 * eq.N, copy=True, tol=tol)
```
Comment (rahulgaur104) [inline tests/test_compute_funs.py:1556]: Maybe add a comment here explaining that the rtol can be reduced if the resolution is increased. Maybe add a TODO link to Kaya's PR and/or comment.

Comment (dpanici) [inline tests/test_compute_funs.py:1556]:
```suggestion
    #TODO: can reduce rtol of test if resolution is increased. See DESC git #1919
    eq_PEST = eq.to_sfl(3 * eq.L, 3 * eq.M, 3 * eq.N, copy=True, tol=tol)
```

**Comment by dpanici** [comment]
Ok no more changes, once tests pass and we confirm they all are passing we can merge

**Review by YigitElma — APPROVED** [review]
